### PR TITLE
Add writer unit tests

### PR DESCRIPTION
**Pull request by EinKrebs**

### CODE REVIEW

File: storage/clickhousedependencystore/dependencystore_test.go [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+import (
+	"context"
+	"testing"
+	"time"
+)
+
+func TestDependencyStore_GetDependencies(t *testing.T) {
+	dependencyStore := DependencyStore{}
+
+	if _, err := dependencyStore.GetDependencies(context.TODO(), time.Now(), time.Hour); err != errNotImplemented {
```
Comment (pavolloffay) [inline storage/clickhousedependencystore/dependencystore_test.go:12]: assert storage to be nil and error to for the specific type.

**Comment by EinKrebs** [comment]
I can't figure out how to unit test span reader and writer, could you please give me some ideas?

**Comment by pavolloffay** [comment]
We might use https://github.com/DATA-DOG/go-sqlmock (or another similar library).

**Comment by EinKrebs** [comment]
Oh, thanks a lot, gonna look at this later.

**Comment by pavolloffay** [comment]
Also please add new make target and add CI job for tests.

**Comment by EinKrebs** [comment]
I don't know, maybe it's better to merge this PR now, before I start to write test for SpanReader, 'cause its size is way too big already?

File: .github/workflows/build.yaml [hunk 2]
```diff
@@ -40,3 +40,18 @@ jobs:
...
+      - uses: actions/checkout@v2.3.4
+        with:
+          submodules: true
+
+      - uses: actions/setup-go@v2
+        with:
+          go-version: ^1.16
+
+      - name: Run unit test
+        run: make unit-tests
```
Comment (pavolloffay) [inline .github/workflows/build.yaml:57]: can we rename it to just `make test`

File: storage/clickhousedependencystore/dependencystore_test.go [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+	"context"
+	"fmt"
+	"testing"
+	"time"
+)
+
+func TestDependencyStore_GetDependencies(t *testing.T) {
+	dependencyStore := DependencyStore{}
+
+	dependencies, err := dependencyStore.GetDependencies(context.TODO(), time.Now(), time.Hour)
```
Comment (pavolloffay) [inline storage/clickhousedependencystore/dependencystore_test.go:13]: use context.Background

File: storage/clickhousedependencystore/dependencystore_test.go [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+	if err != errNotImplemented {
```
Comment (pavolloffay) [inline storage/clickhousedependencystore/dependencystore_test.go:15]: no  if is required, just assert what you expect.

File: storage/clickhousedependencystore/dependencystore_test.go [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+		t.Errorf("Expected GetDependencies not to be implemented, got %s", err)
+	}
+
+	if dependencies != nil {
```
Comment (pavolloffay) [inline storage/clickhousedependencystore/dependencystore_test.go:19]: no if required

File: storage/clickhousespanstore/writer_test.go [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+		OperationName: "operation" + strconv.FormatUint(rand.Uint64(), 10),
+		StartTime:     time.Unix(rand.Int63n(time.Now().Unix()), 0),
+		Process:       &process,
+		Tags:          generateRandomTags(),
+		Duration:      time.Unix(rand.Int63n(1<<32), 0).Sub(time.Unix(0, 0)),
+	}
+	return span
+}
+
+func generateRandomTags() []model.KeyValue {
```
Comment (pavolloffay) [inline storage/clickhousespanstore/writer_test.go:292]: the function is not generating any random tags. 

I would suggest adding a parameter `count` to the signature.

File: storage/clickhousespanstore/writer_test.go [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+		key := "key" + strconv.FormatUint(rand.Uint64(), 16)
+		value := "key" + strconv.FormatUint(rand.Uint64(), 16)
+		kv := model.KeyValue{Key: key, VType: model.ValueType_STRING, VStr: value}
+		tags = append(tags, kv)
+	}
+
+	return tags
+}
+
+func assertExpectationsWereMet(t *testing.T, mock sqlmock.Sqlmock) {
```
Comment (pavolloffay) [inline storage/clickhousespanstore/writer_test.go:304]: not needed just call this in the tests

```assert.NoError(t,  mock.ExpectationsWereMet())```


or 
```
require.NoError(t,  mock.ExpectationsWereMet())
```

if the test continues

File: storage/config_test.go [hunk 8]
```diff
@@ -0,0 +1,27 @@
+package storage
+
+import "testing"
+
+func TestSetDefaults(t *testing.T) {
+	config := Configuration{}
+	config.setDefaults()
+
+	if config.Username != defaultUsername {
```
Comment (pavolloffay) [inline storage/config_test.go:9]: do not use `if`s in this function. directly assert the expected values

File: storage/clickhousespanstore/writer_test.go [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+	testSpansTable = "test_spans_table"
+)
+
+func TestSpanWriter_TagString(t *testing.T) {
+	tags := generateRandomTags()
+	for i := range tags {
+		kv := tags[i]
+		want := fmt.Sprintf("%s=%s", kv.Key, kv.AsString())
+		got := tagString(&kv)
+		if got != want {
```
Comment (pavolloffay) [inline storage/clickhousespanstore/writer_test.go:37]: directly assert on equals

Comment (pavolloffay) [reply storage/clickhousespanstore/writer_test.go:37]: as well in other tests

File: storage/clickhousespanstore/writer_test.go [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+
+const (
+	testSpanCount  = 100
+	testTagCount   = 100
+	testIndexTable = "test_index_table"
+	testSpansTable = "test_spans_table"
+)
+
+func TestSpanWriter_TagString(t *testing.T) {
+	tags := generateRandomTags()
```
Comment (pavolloffay) [inline storage/clickhousespanstore/writer_test.go:32]: Here it might be better to define tags in with all types as a table and test that directly.

Comment (pavolloffay) [reply storage/clickhousespanstore/writer_test.go:32]: https://dave.cheney.net/2019/05/07/prefer-table-driven-tests

File: storage/clickhousespanstore/writer_test.go (hunk 6 above)
Comment (EinKrebs) [reply storage/clickhousespanstore/writer_test.go:292]: > the function is not generating any random tags.

It generates pseudo-random tags, I think it's useful for re-testing.

> I would suggest adding a parameter `count` to the signature.

I'm passing only constant to this parameter, so linter falis and suggests to unparam this.

File: storage/clickhousespanstore/mocks/converter_test.go [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+func TestConverterMock_ConvertValue(t *testing.T) {
+	converter := ConverterMock{}
+
+	type test struct {
+		valueToConvert interface{}
+		expectedResult driver.Value
+	}
+
+	testCases := map[string]test{
+		"string value":       {"some string value", driver.Value("some string value")},
```
Comment (pavolloffay) [inline storage/clickhousespanstore/mocks/converter_test.go:21]: Please use names in struct options. It is recommended good practice.

File: storage/clickhousespanstore/mocks/spylogger_test.go [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+func TestSpyLogger_AssertLogsEmpty(t *testing.T) {
+	logger := NewSpyLogger()
+	logger.AssertLogsEmpty(t)
+}
+
+func TestSpyLogger_AssertLogsOfLevelEqualNoArgs(t *testing.T) {
+	logger := NewSpyLogger()
+	var logs = make([][]LogMock, levelCount)
+	for level, levelLogs := range logs {
+		logsCount := rand.Intn(80)
```
Comment (pavolloffay) [inline storage/clickhousespanstore/mocks/spylogger_test.go:20]: why 80? Magic number?

File: storage/config_test.go [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestSetDefaults(t *testing.T) {
+	config := Configuration{}
+	config.setDefaults()
+
+	assert.Equal(t, config.Username, defaultUsername, "Default username is \"%s\", want \"%s\"", config.Database, defaultUsername)
```
Comment (pavolloffay) [inline storage/config_test.go:13]: I think the config.Username should be as third parameter

File: storage/clickhousespanstore/mocks/converter_test.go (hunk 11 above)
Comment (EinKrebs) [reply storage/clickhousespanstore/mocks/converter_test.go:19]: Can you please give me some guideline for writing table driven tests?

Comment (pavolloffay) [reply storage/clickhousespanstore/mocks/converter_test.go:19]: The test looks good, but the struct definition should list param names.

Comment (EinKrebs) [reply storage/clickhousespanstore/mocks/converter_test.go]: @pavolloffay done

**Review by pavolloffay — APPROVED** [review]

**Pull request merged**